### PR TITLE
radiation advection

### PR DESCRIPTION
**Pull request by chongchonghe**
### Description
Add support for advection on grey radiation.

- Add the O(beta tau) terms to the matter-radiation exchange.
- Add the RadhydroUniformAdvecting test for this new physics. Without O(beta tau), the relative error is about 1e-8 (v = 100 km/s, rho = 1.2, kappa = 100 cm2 g-1). With this update, this test is accurate to machine accuracy.
- Modified boundary condition in RadhydroShock problem
- Passed all tests

### Related issues
N/A

### Checklist

- [x] I have added a description (see above).
- [x] I have added a link to any related issues see (see above).
- [x] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [x] I have added tests for any new physics that this PR adds to the code.
- [x] I have tested this PR on my local computer and all tests pass.
- [x] I have manually triggered the GPU tests with the magic comment `/azp run`.
- [x] I have requested a reviewer for this PR.

### CODE REVIEW

**Comment by BenWibking** [comment]
I see this: "[chongchonghe](https://github.com/chongchonghe) wants to merge 2 commits into [PD-ARS](https://github.com/quokka-astro/quokka/tree/PD-ARS) from [Add-O(beta-tau)-terms](https://github.com/quokka-astro/quokka/tree/Add-O(beta-tau)-terms)"

Do you mean to merge this into the `development` branch instead?

**Comment by BenWibking** [comment]
Also, is this PR only for gray radiation? That's ok -- just want to check.

**Comment by chongchonghe** [comment]
> Also, is this PR only for gray radiation? That's ok -- just want to check.

Yes, advection for grey radiation. I am working on multigroup advection.

**Comment by chongchonghe** [comment]
> I see this: "[chongchonghe](https://github.com/chongchonghe) wants to merge 2 commits into [PD-ARS](https://github.com/quokka-astro/quokka/tree/PD-ARS) from [Add-O(beta-tau)-terms](https://github.com/quokka-astro/quokka/tree/Add-O(beta-tau)-terms)"
> 
> 
> 
> Do you mean to merge this into the `development` branch instead?

I created this PR in vscode, somehow it did that because this branch is based on the PD-ARS branch. Yes, I meant to merge it to dev branch.

**Comment by chongchonghe** [comment]
The GPU tests failed due to running longer than the maximum time of 90 minutes. As I add more and more tests, this will become more of a problem. Perhaps we can remove some of the tests. I removed the RadhydroShockCGS test and reduced the nx of one of the test. 

Ready to merge after review.

**Comment by BenWibking** [comment]
I increased the GPU test timeout to 120 minutes, and re-enabled the RadhydroShockCGS test. Let's see if that works.

File: src/radiation_system.hpp [hunk 1]
```diff
@@ -687,6 +693,108 @@ AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeCellOpticalDepth(const quokka
 						       // return 0.5*(tau_L + tau_R); // arithmetic mean
 }
 
+template <typename problem_t>
+template <FluxDir DIR, typename TARRAY>
+AMREX_GPU_DEVICE void RadSystem<problem_t>::ComputeRadPressure(const double erad, const double Fx, const double Fy, const double Fz, const double fx,
```
Comment (BenWibking) [inline src/radiation_system.hpp]: This should return the values it computes, rather than modify its arguments. It's much clearer that way. There is no performance reason to do otherwise.

File: src/radiation_system.hpp [hunk 2]
```diff
@@ -687,6 +693,108 @@ AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeCellOpticalDepth(const quokka
 						       // return 0.5*(tau_L + tau_R); // arithmetic mean
 }
 
+template <typename problem_t>
+template <FluxDir DIR, typename TARRAY>
```
Comment (BenWibking) [inline src/radiation_system.hpp:697]: Does TARRAY actually need to be a template parameter? It looks like you call this function with the same types every time.

File: src/radiation_system.hpp [hunk 3]
```diff
@@ -687,6 +693,108 @@ AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeCellOpticalDepth(const quokka
...
+	}
+
+	AMREX_ASSERT(Fn != NAN);
+	AMREX_ASSERT(Pnx != NAN);
+	AMREX_ASSERT(Pny != NAN);
+	AMREX_ASSERT(Pnz != NAN);
+
+	F = {Fn, Pnx, Pny, Pnz};
+
+	if (speed_sign == -1) {
```
Comment (BenWibking) [inline src/radiation_system.hpp]: Is this necessary to include in this function? It seems trivial for the caller to do this sign change itself.

File: src/radiation_system.hpp [hunk 4]
```diff
@@ -1108,7 +1101,7 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
 			quokka::valarray<double, nGroups_> deltaErad;
 			quokka::valarray<double, nGroups_> F_R;
 
-			const double resid_tol = 1.0e-13; // 1.0e-15;
+			const double resid_tol = 1.0e-10; // 1.0e-15;
```
Comment (BenWibking) [inline src/radiation_system.hpp]: What is the motivation for this change? This can be made a runtime parameter if it needs to be adjusted per-problem.

File: src/radiation_system.hpp [hunk 5]
```diff
@@ -1206,18 +1198,32 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
...
 
 		amrex::GpuArray<amrex::Real, 3> Frad_t0{};
 		amrex::GpuArray<amrex::Real, 3> Frad_t1{};
 		amrex::GpuArray<amrex::Real, 3> dMomentum{0., 0., 0.};
+		std::array<double, 3> gasMtm{};
+		double realFourPiB = NAN;
 
 		T_gas = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, Egas_guess, massScalars);
-		kappaFVec = ComputeFluxMeanOpacity(rho, T_gas);
+		kappaFVec = ComputeFluxMeanOpacity(rho, T_gas); // note the T_gas is allowed to be NAN
```
Comment (BenWibking) [inline src/radiation_system.hpp:1213]: Why is T_gas allowed to be NAN? That is not immediately obvious to me.

File: tests/RadForce.in [hunk 6]
```diff
@@ -13,7 +13,7 @@ amr.v              = 0       # verbosity in Amr
 # *****************************************************************
 # Resolution and refinement
 # *****************************************************************
-amr.n_cell          = 128 4 4
+amr.n_cell          = 64 4 4
 amr.max_level       = 0     # number of levels = max_level + 1
 amr.blocking_factor = 4     # grid size must be divisible by this
```
Comment (BenWibking) [inline tests/RadForce.in:18]: GPU performance can be improved if the grid parameters are set such that the whole simulation fits in a single box. This should be possible by setting `amr.max_grid_size_x = 128`. Then it may not be necessary to reduce the size of the problem.

File: tests/radshock_dimensionless.in [hunk 7]
```diff
@@ -13,7 +13,7 @@ amr.v              = 0       # verbosity in Amr
 # *****************************************************************
 # Resolution and refinement
 # *****************************************************************
-amr.n_cell          = 512 4 4
+amr.n_cell          = 256 4 4
 amr.max_level       = 0     # number of levels = max_level + 1
 amr.blocking_factor = 4     # grid size must be divisible by this
```
Comment (BenWibking) [inline tests/radshock_dimensionless.in:18]: Same comment as above -- setting `amr.max_grid_size_x = 256` should make this much faster on GPU.

File: src/radiation_system.hpp [hunk 8]
```diff
@@ -32,22 +32,25 @@
...
-
 // SSP-RK2 + implicit radiation-matter exchange
 // static constexpr double IMEX_a22 = 0.0;
 // static constexpr double IMEX_a32 = 0.0;
 
+// physical constants in CGS units
+static constexpr double c_light_cgs_ = C::c_light;	    // cgs
+static constexpr double radiation_constant_cgs_ = C::a_rad; // cgs
+static constexpr double inf = std::numeric_limits<double>::max();
+static constexpr double Erad_zero = 1e-14;
```
Comment (BenWibking) [inline src/radiation_system.hpp]: Is this the radiation energy floor? That should be a runtime parameter if possible.

**Review by BenWibking — CHANGES_REQUESTED** [review]

File: src/radiation_system.hpp (hunk 2 above)
Comment (chongchonghe) [reply src/radiation_system.hpp:697]: It was used as std array and as valarray in two different places.

File: src/radiation_system.hpp (hunk 3 above)
Comment (chongchonghe) [reply src/radiation_system.hpp]: You are right.

File: src/radiation_system.hpp (hunk 5 above)
Comment (BenWibking) [reply src/radiation_system.hpp:1213]: Is this the case for the isothermal EOS?

Comment (chongchonghe) [reply src/radiation_system.hpp:1213]: When gamma_ = 1.0, T_gas is NAN. In this case, we still need to do flux update via the backward Euler method and we need flux-mean opacity for that. The premise is that ComputeFluxMeanOpacity does not take T_gas as a useful parameter.

File: src/radiation_system.hpp (hunk 8 above)
Comment (chongchonghe) [reply src/radiation_system.hpp]: This is the radiation energy fraction floor for each individual group. I'll remove this and replace it with Energy floor in my next PR where I introduce multigroup diffusion test.

Comment (BenWibking) [reply src/radiation_system.hpp]: Can you add a comment in the source explaining this?

File: src/radiation_system.hpp (hunk 2 above)
Comment (chongchonghe) [reply src/radiation_system.hpp:697]: I didn't some changes to the definition and calling of ComputeRadPressure. Now the syntax if slightly more complex but clearer.

File: src/radiation_system.hpp (hunk 4 above)
Comment (chongchonghe) [reply src/radiation_system.hpp]: I just found no benifit is gain from setting a lower resid_tol. Besides, the variable-Kappa advecting pulse problem failed to converged, and I changed this back while trying to resolve it, although it didn't help. I think 1e-10 is more than enough for good conservation.

**Comment by chongchonghe** [comment]
Do I need to do anything for the change request? I have addressed all of them in my recent commits.

**Comment by BenWibking** [comment]
> Do I need to do anything for the change request? I have addressed all of them in my recent commits.

No, just haven't gotten to it yet. I'll try to review this tomorrow my time.

File: src/planck_integral.hpp [hunk 9]
```diff
@@ -243,8 +244,9 @@ AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE auto integrate_planck_from_0_to_x(const
 	if (logx < LOG_X_MIN) {
 		// y = x * x * x / 3.0;    // 1st order
 		y = (-4 + x) * x + 8 * std::log((2 + x) / 2); // 2nd order
-		// AMREX_ASSERT(y <= Y_interp[0]);
-		AMREX_ASSERT(y <= Y_INTERP_MIN);
+		if (y > Y_INTERP_MIN) {
+			y = Y_INTERP_MIN;
+		}
```
Comment (BenWibking) [inline src/planck_integral.hpp:250]: I realized I don't understand what this code is supposed to do. What is `Y_INTERP_MIN` and why does it have a value of `5.13106651231913e-11`?

File: src/radiation_system.hpp [hunk 10]
```diff
@@ -687,6 +693,104 @@ AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeCellOpticalDepth(const quokka
 						       // return 0.5*(tau_L + tau_R); // arithmetic mean
 }
 
+template <typename problem_t>
+template <FluxDir DIR, typename TARRAY>
+AMREX_GPU_DEVICE void RadSystem<problem_t>::ComputeRadPressure(TARRAY &F, double &S, const double erad, const double Fx, const double Fy, const double Fz,
```
Comment (BenWibking) [inline src/radiation_system.hpp:698]: I really think this function should return `std::tuple<TARRAY, double>` instead of `void`.

File: src/radiation_system.hpp [hunk 11]
```diff
@@ -687,6 +693,104 @@ AMREX_GPU_DEVICE auto RadSystem<problem_t>::ComputeCellOpticalDepth(const quokka
...
+	}
+
+	AMREX_ASSERT(Fn != NAN);
+	AMREX_ASSERT(Pnx != NAN);
+	AMREX_ASSERT(Pny != NAN);
+	AMREX_ASSERT(Pnz != NAN);
+
+	F = {Fn, Pnx, Pny, Pnz};
+
+	S = std::max(0.1, std::sqrt(Tnormal));
```
Comment (BenWibking) [inline src/radiation_system.hpp:791]: I really think F and S should be returned rather than modified in place.

File: src/radiation_system.hpp [hunk 12]
```diff
@@ -775,137 +879,29 @@ void RadSystem<problem_t>::ComputeFluxes(array_t &x1Flux_in, array_t &x1FluxDiff
...
+			double S_L = NAN;
+			double S_R = NAN;
+			// ComputeRadPressure will modify F_L and S_L
+			ComputeRadPressure<DIR>(F_L, S_L, erad_L, Fx_L, Fy_L, Fz_L, fx_L, fy_L, fz_L);
+			S_L *= -1.; // speed sign is -1
+				    // ComputeRadPressure will modify F_R and S_R
+			ComputeRadPressure<DIR>(F_R, S_R, erad_R, Fx_R, Fy_R, Fz_R, fx_R, fy_R, fz_R);
+			// speed sign is +1, S_R unchanged.
+
+			// FluxDir::X2
```
Comment (BenWibking) [inline src/radiation_system.hpp]: I think this line is spurious. Can you delete it?

File: src/radiation_system.hpp (hunk 4 above)
Comment (BenWibking) [reply src/radiation_system.hpp]: The issue is that in the optically-think cloud radiation problem I got spurious reflections with a tolerance looser than 1e-12 or so. Can you run that problem and check?

**Review by BenWibking — CHANGES_REQUESTED** [review]

**Comment by BenWibking** [comment]
`RadShadow` is the cloud problem that has problems with spurious reflections unless the relative tolerance is extremely small: https://github.com/quokka-astro/quokka/blob/development/src/RadShadow/test_radiation_shadow.cpp

**Comment by chongchonghe** [comment]
> `RadShadow` is the cloud problem that has problems with spurious reflections unless the relative tolerance is extremely small: https://github.com/quokka-astro/quokka/blob/development/src/RadShadow/test_radiation_shadow.cpp

How to run this test? I enabled this this test and ran, but got this error:

```
Initializing AMReX (23.11-23-g9b733ec45cd9)...
MPI initialized with 1 MPI processes
MPI initialized with thread support level 0
AMReX (23.11-23-g9b733ec45cd9) initialized
radiation epsilon (stiffness parameter) = 1.775153147e-18
Using default ref_ratio = 2 at all levels
Successfully read inputs file ... 
Creating new distribution map on level: 0
Creating new distribution map on level: 1
Creating new distribution map on level: 2
[rsaa-080141:85795] *** Process received signal ***
[rsaa-080141:85795] Signal: Segmentation fault: 11 (11)
[rsaa-080141:85795] Signal code: Invalid permissions (2)
[rsaa-080141:85795] Failing at address: 0x20304062cd3eaf7
[rsaa-080141:85795] [ 0] 0   libsystem_platform.dylib            0x0000000189221a24 _sigtramp + 56
[rsaa-080141:85795] [ 1] 0   ???                                 0x000004062cd3eafe 0x0 + 4424568400638
[rsaa-080141:85795] [ 2] 0   test_radiation_shadow               0x00000001009802d4 _ZN5amrex7AmrMesh12MakeNewGridsEidRiRNS_6VectorINS_8BoxArrayENSt3__19allocatorIS3_EEEE + 2888
[rsaa-080141:85795] [ 3] 0   test_radiation_shadow               0x00000001009818e0 _ZN5amrex7AmrMesh12MakeNewGridsEd + 1592
[rsaa-080141:85795] [ 4] 0   test_radiation_shadow               0x0000000100856cb8 _ZN13AMRSimulationI13ShadowProblemE20setInitialConditionsEv + 88
[rsaa-080141:85795] [ 5] 0   test_radiation_shadow               0x0000000100856b04 _Z12problem_mainv + 252
[rsaa-080141:85795] [ 6] 0   test_radiation_shadow               0x000000010087f41c main + 172
[rsaa-080141:85795] [ 7] 0   dyld                                0x0000000188e710e0 start + 2360
[rsaa-080141:85795] *** End of error message ***
<end of output>
```

**Comment by BenWibking** [comment]
> > `RadShadow` is the cloud problem that has problems with spurious reflections unless the relative tolerance is extremely small: https://github.com/quokka-astro/quokka/blob/development/src/RadShadow/test_radiation_shadow.cpp
> 
> How to run this test? I enabled this this test and ran, but got this error:
> 
> ```
> Initializing AMReX (23.11-23-g9b733ec45cd9)...
> MPI initialized with 1 MPI processes
> MPI initialized with thread support level 0
> AMReX (23.11-23-g9b733ec45cd9) initialized
> radiation epsilon (stiffness parameter) = 1.775153147e-18
> Using default ref_ratio = 2 at all levels
> Successfully read inputs file ... 
> Creating new distribution map on level: 0
> Creating new distribution map on level: 1
> Creating new distribution map on level: 2
> [rsaa-080141:85795] *** Process received signal ***
> [rsaa-080141:85795] Signal: Segmentation fault: 11 (11)
> [rsaa-080141:85795] Signal code: Invalid permissions (2)
> [rsaa-080141:85795] Failing at address: 0x20304062cd3eaf7
> [rsaa-080141:85795] [ 0] 0   libsystem_platform.dylib            0x0000000189221a24 _sigtramp + 56
> [rsaa-080141:85795] [ 1] 0   ???                                 0x000004062cd3eafe 0x0 + 4424568400638
> [rsaa-080141:85795] [ 2] 0   test_radiation_shadow               0x00000001009802d4 _ZN5amrex7AmrMesh12MakeNewGridsEidRiRNS_6VectorINS_8BoxArrayENSt3__19allocatorIS3_EEEE + 2888
> [rsaa-080141:85795] [ 3] 0   test_radiation_shadow               0x00000001009818e0 _ZN5amrex7AmrMesh12MakeNewGridsEd + 1592
> [rsaa-080141:85795] [ 4] 0   test_radiation_shadow               0x0000000100856cb8 _ZN13AMRSimulationI13ShadowProblemE20setInitialConditionsEv + 88
> [rsaa-080141:85795] [ 5] 0   test_radiation_shadow               0x0000000100856b04 _Z12problem_mainv + 252
> [rsaa-080141:85795] [ 6] 0   test_radiation_shadow               0x000000010087f41c main + 172
> [rsaa-080141:85795] [ 7] 0   dyld                                0x0000000188e710e0 start + 2360
> [rsaa-080141:85795] *** End of error message ***
> <end of output>
> ```

See https://github.com/quokka-astro/quokka/pull/476.

**Comment by chongchonghe** [comment]
Since lower resid_tol always gives better results, I'll keep it at 1e-13. I'll try and see if we can pass the RadShadow test after I implement backtracking Newton-Raphson. 

In terms of `ComputeRadPressure` return F and S instead of modifying them at place, I really tried to and couldn't make it work. The problem is std::tuple doesn't work with GPU, and I couldn't find and alternative that works on both CPU and GPU. @BenWibking  Can we proceed with what it is right now? I'll create an issue later and try to make it return F and S.

**Review by BenWibking — APPROVED** [review]
